types/react/experimental.d.ts

```diff
@@ -38,7 +38,24 @@ import React = require('.');
 
 export {};
 
+type NonNullableValue = string | number | boolean | object | symbol;
+
 declare module '.' {
+    type MutableSource<Source extends NonNullableValue, Version extends NonNullableValue> = Pick<{
```

Comment:
It might not be worth it to exclude `null | undefined` but if we want to go that route, the way to do it is this:

```ts
function createMutableSource<Source, Version extends keyof any>(
  source: NonNullable<Source>,
  getVersion: (source: Source) => Version
): MutableSource<Source, Version>
```

This will allow `Source` to be inferred but throw a compilation error when it extends `null | undefined`

Reply:
@ferdaber It would be:

```
// Opaque
type MutableSource<Source, Version> = {
        _source: NonNullable<Source>,
        _getVersion: (source: NonNullable<Source>) => NonNullable<Version>,
};

function createMutableSource<Source, Version>(
        source: NonNullable<Source>,
        getVersion: (source: NonNullable<Source>) => NonNullable<Version>
): MutableSource<Source, Version>;
```
And yeah, this also works as expected.

But constraining types earlier through `extends` here is more intuitive and easier to understand, with less typing and inferences.

Reply:
You won't need to propagate `NonNullable` to non-inference sites like the return type or the `source` parameter in `getVersion`. I'm more partial to having it be this way because it's slightly more future proof until TypeScript has a built in way to explicitly remove types from `any`, but I'm fine with the previous way, it's not a big deal.